### PR TITLE
Use table id as system.sstables partition key

### PR DESCRIPTION
**Pull request by xemul**
The system.sstables (a.k.a. sstables registry) primary key is "string location" as partition key and "uuid generation" as clustering one. The "location" part was taken from table.config.datadir value which, in turn, a string containing path to on-disk files if the table was located locally, e.g. /var/lib/scylla/data/ks/cf-abc123 one. Recently [1] the datadir was moved from table config onto storage options, but this string is still used as registry key.

Other than being owned by a table with ID, sstables are accessed by restore-from-object-storage code [2]. To make it work, both storage driver and sstable_directory helper class maintain two formats of object prefixes for sstables components. For S3-backed sstables having a record in registry, the path used is s3://bucket/generation/component. For restore code there are user-provided prefixes that do not match the aforementioned pattern. The selection between those two is now made by checking sstable state, which is not obvious and may cause troubles for tiered storage driver.

This patch changes  the registry schema so that partition key becomes "uuid owner" and is set to be table.id() value. This is to stop using the local path by S3 backed sstables. Also this change makes it possible for storage driver and sstable directory to rely on the storage options only to tell different bucket prefixes formats from each other.

As a side effect, the make_s3_object_name() helper, that generates the proper object name, becomes explicit for restore-from-S3 usage. Now it relies on the sstable::filename() calling this->prefix() behind the scenes and the latter to return the user-provided prefix, which is pretty fragile construction.

No need to backport (and it's not going to be easy to do it), storage options feature is still experimental

Refs #20675 [1]
Refs #20305 [2]

### CODE REVIEW

**Comment by mykaul** [comment]
> No need to backport (and it's not going to be easy to do it), storage options feature is still experimental

I guess this is the reason there's no feature flag for this, migration, etc?

File: sstables/storage.cc [hunk 1]
```diff
@@ -556,10 +556,15 @@ sstring s3_storage::make_s3_object_name(const sstable& sst, component_type type)
     if (!sst.generation().is_uuid_based()) {
         throw std::runtime_error("'S3' STORAGE only works with uuid_sstable_identifier enabled");
     }
-    if (sst.is_uploaded()) {
-        return format("/{}/{}", _bucket, sst.filename(type));
-    }
-    return format("/{}/{}/{}", _bucket, sst.generation(), sstable_version_constants::get_component_map(sst.get_version()).at(type));
+
```
Comment (tchaikov) [inline sstables/storage.cc:559]: > The s3_storage backend driver has a method that generates object path
withing the bucket

what does "withing" the bucket mean? do you mean "with" or "within"? i searched this word at [w-m](https://www.merriam-webster.com/dictionary/withing), no luck. i think it should be the latter.  just wanted to check with you.

File: db/system_keyspace.cc [hunk 2]
```diff
@@ -3276,36 +3276,36 @@ system_keyspace::read_cdc_generation_opt(utils::UUID id) {
     co_return cdc::topology_description{std::move(entries)};
 }
 
-future<> system_keyspace::sstables_registry_create_entry(utils::UUID location, sstring status, sstables::sstable_state state, sstables::entry_descriptor desc) {
-    static const auto req = format("INSERT INTO system.{} (location, generation, status, state, version, format) VALUES (?, ?, ?, ?, ?, ?)", SSTABLES_REGISTRY);
-    slogger.trace("Inserting {}.{} into {}", location, desc.generation, SSTABLES_REGISTRY);
-    co_await execute_cql(req, location, desc.generation, status, sstables::state_to_dir(state), fmt::to_string(desc.version), fmt::to_string(desc.format)).discard_result();
+future<> system_keyspace::sstables_registry_create_entry(utils::UUID owner, sstring status, sstables::sstable_state state, sstables::entry_descriptor desc) {
```
Comment (tchaikov) [inline db/system_keyspace.cc:3279]: okay, when i was reading the previous commit. was about to ask if we can rename this parameter as well.

**Review by tchaikov — COMMENTED** [review]
lgtm apart from a couple nits.

File: sstables/storage.cc (hunk 1 above)
Comment (xemul) [reply sstables/storage.cc:559]: Should be "within"

**Comment by xemul** [comment]
upd:
- use explicit retyrb type for s3_storage::owner() method
- spelling in patch comment

File: sstables/storage.cc [hunk 3]
```diff
@@ -514,6 +514,10 @@ class s3_storage : public sstables::storage {
 
     sstring make_s3_object_name(const sstable& sst, component_type type) const;
 
+    auto owner() const {
```
[the remark this answers is not in the document]
Comment (xemul) [reply sstables/storage.cc]: Fixed in v2

**Review by tchaikov — APPROVED** [review]
lgtm

File: sstables/storage.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (avikivity) [reply sstables/storage.cc]: It still says "auto"

[the remark this answers is not in the document]
Comment (avikivity) [reply sstables/storage.cc]: Also, I don't see the association between "owner" and "_location".

[the remark this answers is not in the document]
Comment (avikivity) [reply sstables/storage.cc]: ah, I see in a later patch it's not auto

File: db/system_keyspace_sstables_registry.hh [hunk 4]
```diff
@@ -15,23 +15,23 @@ class system_keyspace_sstables_registry : public sstables::sstables_registry {
 public:
     system_keyspace_sstables_registry(system_keyspace& keyspace) : _keyspace(keyspace.shared_from_this()) {}
 
-    virtual seastar::future<> create_entry(sstring location, sstring status, sstables::sstable_state state, sstables::entry_descriptor desc) override {
+    virtual seastar::future<> create_entry(utils::UUID location, sstring status, sstables::sstable_state state, sstables::entry_descriptor desc) override {
```
Comment (avikivity) [inline db/system_keyspace_sstables_registry.hh:18]: Here it should be table_id, which can't be mixed with another uuid. Perhaps in system_keyspace too - it's declared in schema_fwd.hh, so in a fairly low-level module.

File: db/system_keyspace.cc [hunk 5]
```diff
@@ -1120,7 +1120,7 @@ schema_ptr system_keyspace::sstables_registry() {
     static thread_local auto schema = [] {
         auto id = generate_legacy_id(NAME, SSTABLES_REGISTRY);
         return schema_builder(NAME, SSTABLES_REGISTRY, id)
-            .with_column("location", uuid_type, column_kind::partition_key) // will be renamed to "owner" soon
+            .with_column("owner", uuid_type, column_kind::partition_key)
```
Comment (avikivity) [inline db/system_keyspace.cc:1123]: Now I understand "owner" from before a little better.

**Comment by xemul** [comment]
upd:
- use table_id type instead of utils::UUID one
- few more s/location/owner/-s

**Pull request merged**